flake.nix

```diff
@@ -0,0 +1,87 @@
+{
+  inputs = {
+    flake-utils.url = "github:numtide/flake-utils";
+    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+    fenix.url = "github:nix-community/fenix";
+    crane = {
+      url = "github:ipetkov/crane";
+      inputs = {
+        flake-utils.follows = "flake-utils";
+        nixpkgs.follows = "nixpkgs";
+      };
+    };
+  };
+
+  outputs =
+    { self
+    , flake-utils
+    , nixpkgs
+    , fenix
+    , crane
+    ,
+    }:
+    flake-utils.lib.eachDefaultSystem (
+      system:
+      let
+        pkgs = (import nixpkgs) {
+          inherit system;
+          overlays = [
+            fenix.overlays.default
+            (self: super: {
+              rust-dev = super.fenix.stable.withComponents [
+                "cargo"
+                "clippy"
+                "rust-src"
+                "rustc"
+                "rustfmt"
+              ];
+            })
+          ];
+        };
+      in
+      {
+        packages = {
+          default =
+            let
+              craneLib = crane.lib.${system}.overrideToolchain
+                fenix.packages.${system}.minimal.toolchain;
+            in
+            craneLib.buildPackage {
+              pname = "mistralrs-server";
+              buildInputs = with pkgs; [
+                python3
+                pkg-config
+                openssl
+              ];
```

Comment:
@EricLBuehler Essentially, if you add `cargoExtraArgs = "--features ..."` inside the `buildPackage` function, you'd be enabling the features you want. `buildInputs` specifies runtime dependencies, `checkPhaseCargoCommand` runs the tests (it's empty because this gets built in an isolated environment without internet access for reproducibility and the tests want to download stuff), `pname` specifies the package name and the default binary name for `nix run`, and `src` specifies the source code.

Reply:
@EricLBuehler The only way that this can be achieved is by providing a different package definition for each feature so that the user can choose which package they want to build.  This can be done with recursion over a list of features with the only difference between the packages being the feature set. I'm not that good at nix to know how to do it right off the bat but it seems like a cool problem that I might try if you give me a list of features.

Reply:
@EricLBuehler Usually what is done is a function is provided in the nix language where you input the feature set and build the package based on that. This is easier to implement but it doesn't solve the problem of having that accessible through the command line since it's nix.

Reply:
@ivandimitrov8080, can you make this work by having different nix flake files? I'm not familiar with nix though.

The features which I think are important to expose are:
- cuda
- metal
- mkl
- accelerate